### PR TITLE
Add deep energy

### PR DESCRIPTION
**Pull request by tsarikahin**

### CODE REVIEW

File: utils/hyperelasticity/hyperelasticity_utils.py [hunk 1]
```diff
@@ -0,0 +1,364 @@
...
+        # return torch.log(x)
+        return torch.log(torch.maximum(x, torch.tensor(1e-8, dtype=x.dtype, device=x.device)))
+
+def matrix_determinant_2D(a_11, a_22, a_12, a_21):
+    # Calculate the determinant of the 2x2 matrix
+    determinant = a_11 * a_22 - a_12 * a_21
+    return determinant
+
+def matrix_inverse_2D(a_11, a_22, a_12, a_21):
+    # Calculate the determinant
```
Comment (SiVoelkl) [inline utils/hyperelasticity/hyperelasticity_utils.py:100]: `# Calculate the inverse`

File: utils/hyperelasticity/hyperelasticity_utils.py [hunk 2]
```diff
@@ -0,0 +1,364 @@
...
+def matrix_inverse_3D(a11, a12, a13,
+                      a21, a22, a23,
+                      a31, a32, a33):
+    # Compute determinant
+    det = (a11 * (a22 * a33 - a23 * a32)
+         - a12 * (a21 * a33 - a23 * a31)
+         + a13 * (a21 * a32 - a22 * a31))
+    
+    # if det == 0:
+    #     raise ValueError("The matrix is singular and does not have an inverse.")
```
Comment (SiVoelkl) [inline utils/hyperelasticity/hyperelasticity_utils.py:134]: Shouldnt this also be active?

File: utils/hyperelasticity/hyperelasticity_utils.py [hunk 3]
```diff
@@ -0,0 +1,364 @@
...
+    f_xx, f_yy, f_xy, f_yx = deformation_gradient_2D(x, y)
+    
+    # construct C = F^T F
+    C_xx = f_xx * f_xx + f_yx * f_yx
+    C_xy = f_xx * f_xy + f_yx * f_yy
+    C_yx = f_xy * f_xx + f_yy * f_yx
+    C_yy = f_xy * f_xy + f_yy * f_yy
+    
+    if stress_state == "plane_strain":
+        # First invarient
```
Comment (SiVoelkl) [inline utils/hyperelasticity/hyperelasticity_utils.py:193]: `# First invariant`

**Review by SiVoelkl — COMMENTED** [review]
Two minor typos and one change that should be done

File: utils/hyperelasticity/hyperelasticity_utils.py (hunk 1 above)
Comment (danielwolff1) [reply utils/hyperelasticity/hyperelasticity_utils.py:100]:
```suggestion
    # Calculate the inverse
```

File: utils/hyperelasticity/hyperelasticity_utils.py (hunk 3 above)
Comment (danielwolff1) [reply utils/hyperelasticity/hyperelasticity_utils.py:193]:
```suggestion
        # First invariant
```

File: deep_energy_examples/lame/pde_resampler.py
Comment (danielwolff1) [inline deep_energy_examples/lame/pde_resampler.py:1]: This does not look like a lame example to me, is this supposed to be part of this example?

**Review by danielwolff1 — COMMENTED** [review]
Thanks a lot for providing the deep energy code with extensive examples, I believe that this will be a very huge addition to this repo.

I just had some very few minor remarks (off I didn't check all the code in a super detailed fashion, but we also agreed that this is now mostly so that the other developers can see what you're working on, so I do not want you to get caught up in an endless PR discussion 😉)

I just think you have to rebase on `main` before we can merge this because you still use the old build workflow with `setup.py` here and `main` already uses `setup.cfg` but this should be quick and painless. Let me know if there are some problems with that.

Looking forward to having this merged and thanks again for providing the code 🙏

File: utils/hyperelasticity/hyperelasticity_utils.py (hunk 2 above)
Comment (SiVoelkl) [reply utils/hyperelasticity/hyperelasticity_utils.py:134]:
```suggestion
    if det == 0:
        raise ValueError("The matrix is singular and does not have an inverse.")
```

**Comment by SiVoelkl** [comment]
@tsarikahin looks fine to me.
I only wanted to ask if you could provide a short description of the examples as it always takes some time to understand the minimal differences between examples. Thanks

**Comment by tsarikahin** [comment]
Due to time constraints, it would be a bit tricky to write a short description of the examples since there are quite some examples. What I can do is to add a short description to one or two examples, since the idea is the same for other examples as well. I think this should be fine. Indeed, this MR can always be improved in future MRs. Let's be sure that all work is already in the main or master branch.

File: utils/hyperelasticity/hyperelasticity_utils.py (hunk 1 above)
Comment (tsarikahin) [reply utils/hyperelasticity/hyperelasticity_utils.py:100]: Here, first we call the 2d determinant to compute the inverse. This comment defines not the method definition, but what happens below.  In any case, I added a new comment for the matrix inverse part.

**Comment by tsarikahin** [comment]
If it is fine, I suggest meeting next week online to go over some examples to explain what has been done differently based on some examples. Indeed, this meeting should be online :) 

As said earlier, the main differences just lie in the geometry generation and the boundary functions for specifying the external work components in the loss computation. 

Saying this, we can merge the MR if there are no further comments. @danielwolff1 @SiVoelkl

**Review by SiVoelkl — APPROVED** [review]

**Comment by tsarikahin** [comment]
Let's merge it! Sorry guys and thanks a lot, I am super busy with priting and so on, which is  bit tricky. But thanks for the support!!
